File: sbt/sbt-project-highlighting-tests/testdata/projectsForHighlightingTests/local/akka-samples/src/main/java/sample/java/sharding/Device.java
package sample.java.sharding;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

import akka.actor.*;
import akka.event.*;

public class Device extends AbstractActor {

  private LoggingAdapter log = Logging.getLogger(getContext().system(), this);

  private List<Double> temperatures = new ArrayList<Double>();

  public static class RecordTemperature implements Serializable {
    public final Integer deviceId;
    public final Double temperature;

    public RecordTemperature(Integer deviceId, Double temperature) {
      this.deviceId = deviceId;
      this.temperature = temperature;
    }

    @Override
    public String toString() {
      return "RecordTemperature(" + deviceId + ", " + temperature + ")";
    }
  }

  @Override
  public Receive createReceive() {
    return receiveBuilder()
      .match(RecordTemperature.class, r -> {
        temperatures.add(r.temperature);
        log.info("Recording temperature {} for device {}, average is {} after {} readings",
          r.temperature, r.deviceId, sum(temperatures) / temperatures.size(), temperatures.size());
      })
      .build();
  }

  private Double sum(List<Double> doubles) {
    Double result = 0.0;
    for (Double d : doubles) {
      result += d;
    }
    return result;
  }
}

File: sbt/sbt-project-highlighting-tests/testdata/projectsForHighlightingTests/local/akka-samples/src/main/scala/sample/sharding/Device.scala
package sample.sharding

import akka.actor._

/**
 * This is just an example: cluster sharding would be overkill for just keeping a small amount of data,
 * but becomes useful when you have a collection of 'heavy' actors (in terms of processing or state)
 * so you need to distribute them across several nodes.
 */
object Device {
  case class RecordTemperature(deviceId: Int, temperature: Double)
}
class Device extends Actor with ActorLogging {
  import Device._

  override def receive = counting(Nil)

  def counting(values: List[Double]): Receive = {
    case RecordTemperature(id, temp) =>
      val temperatures = temp :: values
      log.info(s"Recording temperature $temp for device $id, average is ${temperatures.sum / temperatures.size} after ${temperatures.size} readings");
      context.become(counting(temperatures))
  }
}
